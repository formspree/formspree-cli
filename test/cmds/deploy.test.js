const deploy = require('@statickit/deploy');
const version = require('../../package.json').version;

jest.mock('process', () => ({
  env: { MY_SECRET: 'pa$$w0rd', API_KEY: '12345' }
}));
jest.mock('@statickit/deploy');
jest.mock('../../src/shim');

jest.mock('ora', () => {
  return () => ({
    start: jest.fn(),
    stop: jest.fn()
  });
});

const cmd = require('../../src/cmds/deploy');
const shim = require('../../src/shim');

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation();
  jest.spyOn(console, 'error').mockImplementation();
});

afterEach(() => {
  console.log.mockRestore();
  console.error.mockRestore();
  shim.install.mockReset();
  deploy.request.mockReset();
});

it('sends a deploy request with the right params', async () => {
  deploy.request.mockImplementation(params => {
    expect(params.config).toStrictEqual({});
    expect(params.key).toBe('xxx');
    expect(params.userAgent).toBe(`@statickit/cli@${version}`);

    return Promise.resolve({ status: 200, data: { id: 'xxxx-xxxx-xxxx' } });
  });

  await cmd.handler({ config: '{}', key: 'xxx' });
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('installs the shim if present in the response', async () => {
  deploy.request.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      data: { id: 'xxxx-xxxx-xxxx', shim: 'shim-ref' }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx', shim: true });
  expect(shim.install).toHaveBeenCalledWith('shim-ref');
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('does not install the shim if not present in the response', async () => {
  deploy.request.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      data: { id: 'xxxx-xxxx-xxxx', shim: null }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx', shim: true });
  expect(shim.install).not.toHaveBeenCalled();
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('does not install the shim if --no-shim is used', async () => {
  deploy.request.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      data: { id: 'xxxx-xxxx-xxxx', shim: 'shim-ref' }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx', shim: false });
  expect(shim.install).not.toHaveBeenCalled();
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('outputs shim install errors', async () => {
  deploy.request.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      data: { id: 'xxxx-xxxx-xxxx', shim: 'shim-ref' }
    });
  });

  shim.install.mockImplementation(async () => {
    throw new Error('error installing package');
  });

  await cmd.handler({ config: '{}', key: 'xxx', shim: true });
  expect(console.error.mock.calls).toMatchSnapshot();
});

it('displays instructions for a missing mailchimp api key', async () => {
  deploy.request.mockImplementation(_params => {
    return Promise.resolve({
      status: 422,
      data: {
        id: 'xxxx-xxxx-xxxx',
        errors: [
          {
            code: 'SECRET_REQUIRED',
            field: 'actions[0].apiKey',
            message: 'is required',
            properties: {
              secret_type: 'mailchimp_api_key',
              secret_key: 'my-mailchimp-key'
            }
          }
        ]
      }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
});

it('displays instructions for a missing mailchimp audience', async () => {
  deploy.request.mockImplementation(_params => {
    return Promise.resolve({
      status: 422,
      data: {
        id: 'xxxx-xxxx-xxxx',
        errors: [
          {
            code: 'SECRET_REQUIRED',
            field: 'actions[0].audience',
            message: 'is required',
            properties: {
              secret_type: 'mailchimp_audience',
              secret_key: 'my-mailchimp-audience'
            }
          }
        ]
      }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
});

it('displays instructions when a secret reference is required', async () => {
  deploy.request.mockImplementation(_params => {
    return Promise.resolve({
      status: 422,
      data: {
        id: 'xxxx-xxxx-xxxx',
        errors: [
          {
            code: 'SECRET_REFERENCE_REQUIRED',
            field: 'actions[0].apiKey',
            message: 'must reference a secret (e.g. @mailchimp-api-key)',
            properties: {
              example_value: 'mailchimp-api-key',
              given_value: 'myinlinekey'
            }
          }
        ]
      }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
});

it('displays general validation errors', async () => {
  deploy.request.mockImplementation(_params => {
    return Promise.resolve({
      status: 422,
      data: {
        id: 'xxxx-xxxx-xxxx',
        errors: [
          {
            code: 'REQUIRED',
            field: 'name',
            message: 'is required',
            properties: {}
          }
        ]
      }
    });
  });

  await cmd.handler({ config: '{}', key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
});

it('substitutes all referenced environment variables', async () => {
  deploy.request.mockImplementation(params => {
    expect(params.config.mySecret).toBe('pa$$w0rd');
    expect(params.config.apiKey).toBe('12345');
    return Promise.resolve({
      status: 200,
      data: { id: 'xxxx-xxxx-xxxx', shim: null }
    });
  });

  const config = {
    mySecret: '$MY_SECRET',
    apiKey: '$API_KEY'
  };

  await cmd.handler({ config: JSON.stringify(config), key: 'xxx' });
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('throws an error if undefined env vars are referenced', async () => {
  const config = {
    mySecret: '$MY_SECRET_1',
    apiKey: '$API_KEY_1'
  };

  await cmd.handler({ config: JSON.stringify(config), key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
});

it('throws an error if apiKey properties do not point to env vars', async () => {
  const config = {
    forms: {
      contactForm: {
        actions: [
          {
            apiKey: 'my-inline-key'
          }
        ]
      }
    }
  };

  await cmd.handler({ config: JSON.stringify(config), key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('throws an error if apiSecret properties do not point to env vars', async () => {
  const config = {
    forms: {
      contactForm: {
        actions: [
          {
            apiSecret: 'my-inline-key'
          }
        ]
      }
    }
  };

  await cmd.handler({ config: JSON.stringify(config), key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('throws an error if secretKey properties do not point to env vars', async () => {
  const config = {
    forms: {
      contactForm: {
        actions: [
          {
            secretKey: 'my-inline-key'
          }
        ]
      }
    }
  };

  await cmd.handler({ config: JSON.stringify(config), key: 'xxx' });
  expect(console.error.mock.calls).toMatchSnapshot();
  expect(console.log.mock.calls).toMatchSnapshot();
});

it('skips validating inline secrets with force flag', async () => {
  const config = {
    apiKey: 'my-inline-key'
  };

  deploy.request.mockImplementation(params => {
    expect(params.config.apiKey).toBe('my-inline-key');
    return Promise.resolve({
      status: 200,
      data: { id: 'xxxx-xxxx-xxxx', shim: null }
    });
  });

  await cmd.handler({
    config: JSON.stringify(config),
    key: 'xxx',
    force: true
  });
  expect(console.log.mock.calls).toMatchSnapshot();
});
