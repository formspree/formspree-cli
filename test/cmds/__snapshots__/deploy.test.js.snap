// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays general validation errors 1`] = `
Array [
  Array [
    "[31m[1mâœ•[22m[39m [31m[1mDeployment failed [90m(xxxx-xxxx-xxxx)[31m[22m[39m",
  ],
  Array [
    "",
  ],
  Array [
    "  1) [36mname[39m is required",
  ],
  Array [
    "",
  ],
]
`;

exports[`displays instructions for a missing mailchimp api key 1`] = `
Array [
  Array [
    "[31m[1mâœ•[22m[39m [31m[1mDeployment failed [90m(xxxx-xxxx-xxxx)[31m[22m[39m",
  ],
  Array [
    "",
  ],
  Array [
    "  1) The secret [36m[1m@my-mailchimp-key[22m[39m is not defined yet [90m(actions[0].apiKey)[39m",
  ],
  Array [
    "",
  ],
  Array [
    "       Here's where to find your API key:
       
         â€¢ Log in to your Mailchimp account
         â€¢ Open the menu with your avatar in the upper right
         â€¢ Navigate to [1m[36mProfile â€º Extras â€º API keys[39m[22m
         â€¢ Under \\"Your API keys\\", click [1m[36mCreate A Key[39m[22m
       
       Copy the generated key and run the following command:
       
         [90m$[39m statickit secrets add [36mmy-mailchimp-key[39m [33m<paste-api-key-here>[39m",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
]
`;

exports[`displays instructions for a missing mailchimp audience 1`] = `
Array [
  Array [
    "[31m[1mâœ•[22m[39m [31m[1mDeployment failed [90m(xxxx-xxxx-xxxx)[31m[22m[39m",
  ],
  Array [
    "",
  ],
  Array [
    "  1) The secret [36m[1m@my-mailchimp-audience[22m[39m is not defined yet [90m(actions[0].audience)[39m",
  ],
  Array [
    "",
  ],
  Array [
    "       Here's how to found your Audience ID:
       
         â€¢ Log in to your Mailchimp account
         â€¢ Click the [1m[36mAudience[39m[22m tab at the top
         â€¢ Navigate to [1m[36mManage Audience â€º Settings[39m[22m
         â€¢ Click [1m[36mAudience name and defaults[39m[22m
         â€¢ Locate the value under the \\"Audience ID\\" heading
       
       Copy the ID and run the following command:
       
         [90m$[39m statickit secrets add [36mmy-mailchimp-audience[39m [33m<paste-audience-id-here>[39m",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
]
`;

exports[`displays instructions when a secret reference is required 1`] = `
Array [
  Array [
    "[31m[1mâœ•[22m[39m [31m[1mDeployment failed [90m(xxxx-xxxx-xxxx)[31m[22m[39m",
  ],
  Array [
    "",
  ],
  Array [
    "  1) [36m[1mactions[0].apiKey[22m[39m must reference a secret (e.g. [36m[1m@mailchimp-api-key[22m[39m)",
  ],
  Array [
    "",
  ],
  Array [
    "       This field must reference a secret instead of a raw value, 
       because it's too sensitive to commit to version control.
       
       [33m[1m-- Instructions ----------------------------------------------------------[22m[39m
       
       First, choose a name for this value and run the following command:
       
           [90m$[39m statickit secrets add [36mmailchimp-api-key[39m [32m\\"myinlinekey\\"[39m
       
       Then, reference your secret like this (notice the @-symbol prefix):
       
           {
             [32m\\"apiKey\\"[39m: [32m\\"@mailchimp-api-key\\"[39m
           }",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
]
`;

exports[`does not install the shim if --no-shim is used 1`] = `
Array [
  Array [
    "[32mâœ”[39m [32mDeployment succeeded [90m(xxxx-xxxx-xxxx)[32m[39m",
  ],
]
`;

exports[`does not install the shim if not present in the response 1`] = `
Array [
  Array [
    "[32mâœ”[39m [32mDeployment succeeded [90m(xxxx-xxxx-xxxx)[32m[39m",
  ],
]
`;

exports[`installs the shim if present in the response 1`] = `
Array [
  Array [
    "[32mâœ”[39m [32mDeployment succeeded [90m(xxxx-xxxx-xxxx)[32m[39m",
  ],
  Array [
    "[32mâœ”[39m [32mFunctions installed [90m(shim-ref)[32m[39m",
  ],
]
`;

exports[`outputs shim install errors 1`] = `
Array [
  Array [
    "[31m[1mâœ•[22m[39m [31m[1mFunctions failed to install[22m[39m",
  ],
  Array [
    [Error: error installing package],
  ],
]
`;

exports[`sends a deploy request with the right params 1`] = `
Array [
  Array [
    "[32mâœ”[39m [32mDeployment succeeded [90m(xxxx-xxxx-xxxx)[32m[39m",
  ],
]
`;

exports[`substitutes all referenced environment variables 1`] = `
Array [
  Array [
    "[32mâœ”[39m [32mDeployment succeeded [90m(xxxx-xxxx-xxxx)[32m[39m",
  ],
]
`;

exports[`throws an error if undefined env vars are referenced 1`] = `
Array [
  Array [
    "[31m[1mâœ•[22m[39m [31m[1mThe following environment variables were referenced but are not defined: MY_SECRET_1, API_KEY_1[22m[39m",
  ],
]
`;
